### PR TITLE
Enhance Error Handling in ZIMArchive and ZIMArchiveLoader | issue #813

### PR DESCRIPTION
**Pull request by ombalgude**
We have successfully refactored ZIMArchive and ZIMArchiveLoader to eliminate UI dependencies, thereby enhancing backend adaptability. Instead of using direct UI alerts, we have implemented callback-based error handling, which enables more efficient error management. Furthermore, we have updated methods to ensure improved error reporting and modularity. This modification increases compatibility with environments such as Web Workers and optimizes overall error management.

### CODE REVIEW

**Review by ombalgude — COMMENTED** [review]
helpfully👍

**Comment by ombalgude** [comment]
helpfully👍

**Comment by ombalgude** [comment]
This is my first open-source contribution. I hope it will be helpful. I enjoyed contributing. If there are any changes, please suggest them.😄

File: www/js/lib/zimArchive.js [hunk 1]
```diff
@@ -98,20 +108,132 @@ function ZIMArchive (storage, path, callbackReady, callbackError) {
...
+        } else {
+            // All listings should be loaded, so we can now call the callback
+            if (callbackReady) {
+                callbackReady({ success: true, archive: that });
             }
-        });
+        }
+    })
+    .catch(handleError); // Handle errors during initial metadata addition
+};
```
Comment (audiodude) [inline www/js/lib/zimArchive.js:132]: This looks like an error to me. This brace closes:

`function ZIMArchive (storage, path, callbackReady, callbackError) {`

And therefore orphans all the code underneath it. Please fix.

**Review by audiodude — CHANGES_REQUESTED** [review]
Thank you so much for the code contribution! I like the general idea of this code, and I look forward to reviewing it further but I got caught up because it looks like it may be broken. Please take a look and let me know.

Also, please refrain from adding unrelated style changes (whitespace including indentation, parameter spacing, single quotes -> double quotes, etc)

File: www/js/lib/zimArchive.js (hunk 1 above)
Comment (ombalgude) [reply www/js/lib/zimArchive.js:132]: Thank you for the feedback. I'll check it out and make correction

**Review by ombalgude — COMMENTED** [review]
@audiodude I've made the requested changes. Please review them when you have a moment.

**Comment by Jaifroid** [comment]
@ombalgude Thank you for your contribution! We appreciate your effort to improve the project. To help us better understand and review your changes, could you please provide some additional information?

1. Which issue is this PR addressing? If there isn't an existing issue, it would be helpful to create one explaining the rationale behind these changes.
2. Could you clarify which dependencies you were aiming to eliminate?
3. We noticed several changes without accompanying comments or explanations. It would be great if you could add some context.
4. You've made some stylistic changes (e.g., switching from single to double quotes), which we do warn against doing in [CONTRIBUTING](https://github.com/kiwix/kiwix-js/blob/main/CONTRIBUTING.md), so kindly go through and eliminate purely stylistic changes, so we can see what is of substance in your PR.

For future contributions, here are some tips that can help streamline the process:

* Discuss your proposed approach on the relevant issue before submitting a PR. This allows for valuable feedback early in the process.
* For style-related changes, please refer to our [CONTRIBUTING](https://github.com/kiwix/kiwix-js/blob/main/CONTRIBUTING.md) guide. Following these guidelines ensures consistency across the project. We have an ESLint definition that can ensure you stick to the house style if you use an IDE that supports it.
* When requesting a review, it's helpful to include information about the testing you've performed.

If you have any questions about these suggestions, please don't hesitate to ask.

**Comment by Jaifroid** [comment]
@ombalgude I'm afraid tests are still failing. To diagnose the issue, please look at https://github.com/kiwix/kiwix-js/actions/runs/10064097481/job/27827615748?pr=1273 .

**Comment by ombalgude** [comment]
> Please see https://github.com/kiwix/kiwix-js/blob/main/CONTRIBUTING.md#testing and https://github.com/kiwix/kiwix-js/blob/main/TESTS.md to understand the automatic tests we do. The info should help you to work with the tests and debug your PR. Thanks.

@Jaifroid you please provide some guidance on the modifications required to successfully merge this PR? Any hints would be greatly appreciated.

**Comment by Jaifroid** [comment]
> @Jaifroid you please provide some guidance on the modifications required to successfully merge this PR? Any hints would be greatly appreciated.

@ombalgude First of all I think you need to explain to us what this PR is aiming to do precisely. The corresponding issue had been fixed already, but it was reopened in 2022 because some unfixed dependency on uiUtil.js was found in dec_wrapper.js. I was slightly surprised to see your code doesn't touch dec_wrapper.js. It doesn't mean what you've done is invalid, but the main problem is that we have no idea what you actually changed in zimArchive.js or zimArchiveLoader.js due to the many many style changes. So, you need to start by reverting all the style changes. Until you do that we can't really help!

**Comment by ombalgude** [comment]
Hey @Jaifroid , I'm done with the new changes. Please check them out and give me some valuable feedback.

File: CONTRIBUTING.md
Comment (Jaifroid) [inline CONTRIBUTING.md:1]: It seems to me that this file has nothing to do with the PR, and shouldn't be part of it.

File: TESTS.md
Comment (Jaifroid) [inline TESTS.md:1]: Ditto for this file. Kindly avoid changing files that are not directly related to the issue you are fixing.

File: www/js/lib/zimArchive.js [hunk 2]
```diff
@@ -20,16 +20,16 @@
...
-import util from './util.js';
-import uiUtil from './uiUtil.js';
-import utf8 from './utf8.js';
-import translateUI from './translateUI.js';
+import zimfile from "./zimfile.js";
+import zimDirEntry from "./zimDirEntry.js";
+import util from "./util.js";
+import uiUtil from "./uiUtil.js";
+import utf8 from "./utf8.js";
+import translateUI from "./translateUI.js";
```
Comment (Jaifroid) [inline www/js/lib/zimArchive.js:32]: These are all stylistic changes and shouldn't be in the PR.

File: www/js/lib/zimArchive.js [hunk 3]
```diff
@@ -71,153 +71,117 @@ var LZ;
...
-            });
-        });
-    };
-    if (storage && !path) {
-        var fileList = storage;
-        // We need to convert the FileList into an Array
-        var fileArray = [].slice.call(fileList);
-        // The constructor has been called with an array of File/Blob parameter
-        createZimfile(fileArray);
+function ZIMArchive(storage, path, callbackReady, callbackError) {
```
Comment (Jaifroid) [inline www/js/lib/zimArchive.js:74]: What concrete code changes have you made to the ZIMArchive object definition?  There are a mass of changes that are not explained, and it's impossible for me to see concretely what has changed in substance.

File: www/js/lib/zimArchive.js [hunk 4]
```diff
@@ -256,34 +231,165 @@ ZIMArchive.prototype.isReady = function () {
...
+          params.zimType = that.setZimType();
+          // If user is not using libzim for reading the file, we can call the ready callback now
+          if (!params.useLibzim) whenZimReady();
+        });
+      })
+      .catch(function (err) {
+        console.warn("Error setting archive listings: ", err);
+      });
+  });
+  {
```
Comment (Jaifroid) [inline www/js/lib/zimArchive.js:387]: Why have you added lines 260 to 387?

File: www/js/lib/zimArchive.js
Comment (Jaifroid) [inline www/js/lib/zimArchive.js:1]: Changes in this file are illegible, and many are still style changes, or indentation changes.

File: www/js/lib/zimArchiveLoader.js
Comment (Jaifroid) [inline www/js/lib/zimArchiveLoader.js:1]: Ditto for this file.

**Review by Jaifroid — CHANGES_REQUESTED** [review]
Unfortunately, I don't see any difference with your latest commit. There are still a bunch of style and indentation changes that make the PR illegible. Please follow the previously given advice carefully, so we can understand what this PR attempts to do.

**Comment by Jaifroid** [comment]
@ombalgude Tests are still failing. You don't appear to have followed our advice to remove all stylistic changes. Most of your changes seem to be moving code around, but it's not clear what you've fixed.

I did suggest above that you please discuss and explain what your PR does in writing, and you have not done this.

Also, please update the branch from main (there is a button "Update branch" below, that should make this easy to do).

**Comment by ombalgude** [comment]
Ok @Jaifroid

**Comment by Jaifroid** [comment]
@ombalgude Do you intend to keep working on this PR, or shall I close it? Thanks.

**Comment by Jaifroid** [comment]
So I'm closing, but feel free to re-open if there are further contributions.

**Comment by BirushaNdegeya** [comment]
I am going to work on this task. If you have attempted to solve this issue before, I would appreciate it if you could share what you have done so far. It would help me as a beginner in this project.
